src/main/java/com/google/api/codegen/ApiaryConfig.java

```diff
@@ -102,15 +102,59 @@
   private final Table<Type, String, Field> fields = HashBasedTable.<Type, String, Field>create();
 
   /*
-   * The service canonical name, or name if no canonical name
+   * The service canonical name, or name if no canonical name.
    */
   private String serviceCanonicalName;
 
   /*
-   * The service version string
+   * The service version string.
    */
   private String serviceVersion;
 
+  private Map<String, AuthType> authOverrides = new HashMap<>();
+
+  /*
+   * If present in the scope list, indicates that the API supports application default credentials
+   * based auth.
+   */
+  private static final String CLOUD_PLATFORM_SCOPE =
+      "https://www.googleapis.com/auth/cloud-platform";
+
+  /*
+   * Possible auth types supported by discovery.
+   */
+  public enum AuthType {
+    APPLICATION_DEFAULT_CREDENTIALS,
+    OAUTH_3L,
+    API_KEY
+  }
+
+  /*
+   * Returns the auth type supported by the service.
+   */
+  public AuthType getAuthType() {
+    String key = getServiceCanonicalName();
+    if (authOverrides.containsKey(key)) {
+      return authOverrides.get(key);
+    }
+    // This statement is based on the assumption that every method in a service contains all the
```

Comment:
Actually I'd really like to leave it in the service form... but I realized a slightly more robust way to do it:

If the API doesn't have a single scope (`getAuthScopes().isEmpty()`), then we know it's API-key based. If it has more than one scope, and one of the scopes is `"https://googleapis.com/auth/cloud-platform"`, then we know it's ADC based. Otherwise it's 3LO.

I don't think this can fail in the context of discovery, and since it doesn't change how code is generated down the line, none of the `readonly` or scope specific semantics makes a difference.